drivers/filter/admv8818/admv8818.c

```diff
@@ -0,0 +1,423 @@
+	{12550000000, 18500000000}
+};
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * @brief Writes data to ADMV8818 over SPI.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data value to write.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv8818_spi_write(struct admv8818_dev *dev, uint8_t reg_addr,
+		       uint8_t data)
+{
+	uint8_t buff[ADMV8818_BUFF_SIZE_BYTES];
+
+	buff[0] = reg_addr;
+	buff[1] = data;
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff,
+					ADMV8818_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Reads data from ADMV8818 over SPI.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data read from the device.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv8818_spi_read(struct admv8818_dev *dev, uint8_t reg_addr,
+		      uint8_t *data)
+{
+	uint8_t buff[ADMV8818_BUFF_SIZE_BYTES];
+	int ret;
+
+	buff[0] = ADMV8818_SPI_READ_CMD | reg_addr;
+	buff[1] = 0;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff, ADMV8818_BUFF_SIZE_BYTES);
+	if(ret)
+		return ret;
+
+	*data = buff[1];
+
+	return 0;
+}
+
+/**
+ * @brief Update ADMV8818 register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Mask for specific register bits to be updated.
+ * @param data - Data written to the device (requires prior bit shifting).
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv8818_spi_update_bits(struct admv8818_dev *dev, uint8_t reg_addr,
+			     uint8_t mask, uint8_t data)
+{
+	uint8_t read_val;
+	int ret;
+
+	ret = admv8818_spi_read(dev, reg_addr, &read_val);
+	if (ret)
+		return ret;
+
+	read_val &= ~mask;
+	read_val |= data;
+
+	return admv8818_spi_write(dev, reg_addr, read_val);
+}
+
+/**
+ * @brief Set the HPF Frequency.
+ * @param dev - The device structure.
+ * @param freq - The HPF Frequency to be set.
+ * @return Returns 0 in case of success or negative error code.
+ */
+int admv8818_hpf_select(struct admv8818_dev *dev, unsigned long long freq)
+{
+	unsigned int hpf_step = 0, hpf_band = 0, i, j;
+	unsigned long long freq_step;
+	int ret;
+
+	if (freq < freq_range_hpf[0][0])
+		goto hpf_write;
+
+	if (freq > freq_range_hpf[3][1]) {
+		hpf_step = 15;
+		hpf_band = 4;
+
+		goto hpf_write;
+	}
+
+	for (i = 0; i < 4; i++) {
+		freq_step = (freq_range_hpf[i][1] - freq_range_hpf[i][0]) / 15;
+
+		if (freq > freq_range_hpf[i][0] &&
+		    (freq < freq_range_hpf[i][1] + freq_step)) {
+			hpf_band = i + 1;
+
+			for (j = 1; j <= 16; j++) {
+				if (freq < (freq_range_hpf[i][0] + (freq_step * j))) {
+					hpf_step = j - 1;
+					break;
+				}
+			}
+			break;
+		}
+	}
```

Comment:
nitpick: would it be useful to add a comment here which explains these operations?

Reply:
Sure, I can add, but we usually add comments only is special cases, like datasheet not public / particular uncommon operation. In this case, and in most of the other cases the datasheet is bound to the commit description of the driver implementation so that the reviewers can analyze the code based on the documentation available (https://github.com/torvalds/linux/commit/f34fe888ad0546dacf678aa604435d442934984f). And if there are discrepancies between the code and the datasheet documentation, then we should intervene with inline comments.